appcues/src/main/java/com/appcues/session/DefaultSessionService.kt

```diff
@@ -0,0 +1,176 @@
+            userSignature = sessionLocalSource.getUserSignature(),
+            latestUserProperties = latestUserProperties,
+            properties = hashMapOf(
+                "userId" to userId,
+                "_sessionId" to sessionId,
+                "_isAnonymous" to sessionLocalSource.isAnonymous(),
+                "_localId" to sessionLocalSource.getDeviceId(),
+                "_lastContentShownAt" to sessionLocalSource.getLastContentShownAt(),
+                "_currentScreenTitle" to currentScreen,
+                "_lastScreenTitle" to previousScreen,
+                "_sessionPageviews" to sessionPageviews,
+                "_sessionRandomizer" to sessionRandomId,
+            ).filterValues { it != null }.mapValues { it.value as Any }
+        )
+    }
+
+    override suspend fun checkSession(intent: AnalyticsIntent, onSessionStarted: suspend () -> Unit): Boolean {
+        // the order of conditions here is important
+        return when {
+            // check if the incoming intent is Anonymous or Identify, we force start a new session
+            intent is Anonymous -> startAnonymousSession().also { onSessionStarted() }
+            intent is Identify -> startSession(intent).also { onSessionStarted() }
+            // for any other intent types we check if session is started already
+            isSessionStarted() -> true
+            // if not we try to start a session based on stored information
+            startSession() -> true.also { onSessionStarted() }
+            // we could not start or check for existing session
+            else -> false
+        }.updateSession(intent)
+    }
+
+    private fun Boolean.updateSession(intent: AnalyticsIntent): Boolean {
+        if (this) {
+            when {
+                intent is Anonymous -> {
+                    latestUserProperties = hashMapOf()
+                }
+                intent is Identify -> {
+                    latestUserProperties = intent.properties.sanitized()
+                }
+                intent is UpdateProfile -> {
+                    latestUserProperties = intent.properties.sanitized()
+                }
+                intent is Screen -> {
+                    previousScreen = currentScreen
+                    currentScreen = intent.title
+                    sessionPageviews += 1
+                }
+                intent is Event && intent.name == AnalyticsEvent.SessionStarted.eventName -> {
+                    // special handling for session start events
+                    sessionPageviews = 0
+                    sessionRandomId = sessionRandomizer.get()
+                    currentScreen = null
+                    previousScreen = null
+                }
+            }
+        }
+
+        return this
+    }
+
+    private fun Map<String, Any>?.sanitized(): Map<String, Any> {
+        return mutableMapOf<String, Any>().apply {
+            putAll(this@sanitized ?: hashMapOf())
+            remove(KEY_USER_SIGNATURE)
```

Comment:
its very convenient that we "access" and also "clear" the user signature here in SessionService. dont need to about this property anywhere else.